trio/_core/_tests/test_instrumentation.py

```diff
@@ -1,40 +1,45 @@
+from __future__ import annotations
+
+from typing import Container, Iterable, NoReturn
+
 import attr
 import pytest
 
 from ... import _abc, _core
+from ...lowlevel import Task
 from .tutil import check_sequence_matches
 
 
 @attr.s(eq=False, hash=False)
-class TaskRecorder:
-    record = attr.ib(factory=list)
+class TaskRecorder(_abc.Instrument):
+    record: list[tuple[str, Task | None]] = attr.ib(factory=list)
 
-    def before_run(self):
-        self.record.append(("before_run",))
+    def before_run(self) -> None:
+        self.record.append(("before_run", None))
 
-    def task_scheduled(self, task):
+    def task_scheduled(self, task: Task) -> None:
         self.record.append(("schedule", task))
 
-    def before_task_step(self, task):
+    def before_task_step(self, task: Task) -> None:
         assert task is _core.current_task()
         self.record.append(("before", task))
 
-    def after_task_step(self, task):
+    def after_task_step(self, task: Task) -> None:
         assert task is _core.current_task()
         self.record.append(("after", task))
 
-    def after_run(self):
-        self.record.append(("after_run",))
+    def after_run(self) -> None:
+        self.record.append(("after_run", None))
 
-    def filter_tasks(self, tasks):
+    def filter_tasks(self, tasks: Container[Task]) -> Iterable[tuple[str, Task | None]]:
```

Comment:
Doesn't matter a whole lot, but I think `Iterable` is best used in return type if it's returning an object that defines an `__iter__` method. Even though it's more text, I think `Generator` would be more applicable here.

Reply:
I prefer to generally use `Iterator`, since it leaves you open to changing the returned object to something else in the future. The fact it's a generator is an implementation detail. Unless you do want the users to use generator-specific methods.

Reply:
Yep for `contextlib.contextmanager` it needs to be a generator, since that decorator uses `.send()`/`.throw()`. The type stubs are wrong, but changing them would require too much churn...